### PR TITLE
Wavefront Opentracing SDK Ruby

### PR DESCRIPTION
**Pull request by gangadharaswamy**
Wavefront Opentracing SDK Ruby tested on console reporter. Receiving data as below:

`"test_op" source="unknown" traceId=080c6a6c-a3f8-45a2-a24f-402753b3fec5 spanId=080c6a6c-a3f8-45a2-a24f-402753b3fec5 "key1"="val1" "key1"="val2" "application"="app" "service"="service" "cluster"="us-west-1" "shard"="primary" "custom_k"="custom_v" 1549663388000 0`

Testing is performed using MiniTest gem.
Code coverage is above 80%. Rake test files are added.

### CODE REVIEW

File: lib/wavefrontopentracing/propagation/propagator.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+# @author: Gangadharaswamy (gangadhar@vmware.com)
+
+
+module Propagation
+
+  class Propagator
+
+    def inject(_span_context, _carrier)
+      # Inject the given context into the given carrier.
+      # @param span_context [SpanContext]: The span context to serialize
```
Comment (yogeshprasad) [inline lib/wavefrontopentracing/propagation/propagator.rb]: Follow the YARD guideline for comments (Check all the files)
https://www.rubydoc.info/gems/yard/file/docs/GettingStarted.md 
and this as well

https://github.com/rubocop-hq/ruby-style-guide

File: lib/wavefrontopentracing/reporting/composite.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    end
+
+    def failure_count
+      # Total failure count of all reporters
+      # @return [int]: Total failure count
+      total_count = 0
+      @reporters.each do |rep|
+        total_count += rep.failure_count
+      end
+      reporter
```
Comment (yogeshprasad) [inline lib/wavefrontopentracing/reporting/composite.rb]: return total_count

File: util/utils.rb [hunk 3]
```diff
@@ -0,0 +1,188 @@
+# Utils module contains useful function for preparing and processing data.
+#
+# @author Yogesh Prasad Kurmi (ykurmi@vmware.com)
+
+require "zlib"
+require 'scanf'
+
```
Comment (yogeshprasad) [inline util/utils.rb]: Please remove all tier one SDK files.

File: lib/wavefrontopentracing/propagation/textmap.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      end
+
+      carrier.each do |key, val|
+        key = key.downcase
+        if key.include? TRACE_ID
+          trace_id = val # To-Do: if validate_uuid(val)
+        elsif key.include? SPAN_ID
+          span_id = val # To-Do: if validate_uuid(val)
+        elsif key.include? BAGGAGE_PREFIX
+          baggage.merge!(key[BAGGAGE_PREFIX.length..-1] => val)
```
Comment (yogeshprasad) [inline lib/wavefrontopentracing/propagation/textmap.rb]: please check this.

**Review by yogeshprasad — CHANGES_REQUESTED** [review]

File: lib/wavefrontopentracing/propagation/registry.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      # Get propagator of certain format.
+      # @param format [OpenTracing::FORMAT] : Format a propagator
+      # @return Propagator of given format
+
+      @propagators[format]
+    end
+
+    def register(format, propagator)
+      # Register propagator.
+      # @param format [OpenTracing::FORMAT] : Format a propagator
```
Comment (yogeshprasad) [inline lib/wavefrontopentracing/propagation/registry.rb]: : can be removed

File: lib/wavefrontopentracing/propagation/registry.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      # @param propagator [Propagator] : Propagator to be registered.
```
Comment (yogeshprasad) [inline lib/wavefrontopentracing/propagation/registry.rb]: : can be removed

File: wavefront-opentracing-sdk-ruby.gemspec [hunk 7]
```diff
@@ -0,0 +1,32 @@
+lib = File.expand_path('lib', __dir__)
+$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
+require_relative 'lib/wavefrontopentracing/version'
+
+Gem::Specification.new do |spec|
+  spec.name = 'wavefront-opentracing-sdk'
+  spec.version = WavefrontOpentracing::VERSION
+  spec.authors = ['wavefront-opentracing-sdk']
```
Comment (yogeshprasad) [inline wavefront-opentracing-sdk-ruby.gemspec]: please update valid authors

File: test/reporter_test.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  def self.app_tags
+    Wavefront::ApplicationTags.new("wft2-shopping",
+                                   "shopping-service",
+                                   cluster: "virunga",
+                                   shard: "shopping-shard",
+                                   custom_tags: {"size" => "XL", "color" => "black"})
+  end
+
+  def test_composite_reporter
+    proxy_host = '10.192.167.61'
```
Comment (yogeshprasad) [inline test/reporter_test.rb]: remove IP

File: test/reporter_test.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    #   reporter and console reporter. Report span data to all the reporters.
+    proxy_client = Wavefront::WavefrontProxyClient.new(proxy_host,
+                                                       metrics_port,
+                                                       distribution_port,
+                                                       tracing_port)
+    preporter = Reporting::WavefrontSpanReporter.new(client: proxy_client,
+                                                       source: "proxy")
+
+    direct_client = Wavefront::WavefrontDirectIngestionClient.new(
+                      'https://nimba.wavefront.com',
```
Comment (yogeshprasad) [inline test/reporter_test.rb]: remove the API key and cluster name

File: Gemfile.lock [hunk 10]
```diff
@@ -0,0 +1,61 @@
+PATH
```
Comment (yogeshprasad) [inline Gemfile.lock]: please delete this file

**Review by yogeshprasad — CHANGES_REQUESTED** [review]
some cosmetic changes,
Otherwise LGTM

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+before_install:
+  - gem install test-unit
+  - gem install bundler -v 2.0.1
+  - gem install simplecov -v 0.16.1
+  - gem install simplecov-console -v 0.4.2
+  - gem install concurrent-ruby
+  - gem install opentracing -v 0.5.0
+  - gem install bump
+  - gem install minitest
+#  - gem install wavefront-client
```
Comment (sushantdewan123) [inline .travis.yml]: please cleanup (delete) stale code instead of commenting out.

File: README.md [hunk 12]
```diff
@@ -1 +1,128 @@
...
+require 'wavefrontopentracing'
+require 'wavefront/client/direct'
+# or
+# require 'wavefront/client/proxy'
+
+# Wavefront sender for direct ingestion
+wavefront_sender = Wavefront::WavefrontDirectIngestionClient.new('<wavefront-cluster>', '<API-Token>')
+# or
+# Wavefront sender for proxy
+# default values: metrics_port = 2878, distribution_port = 40000, tracing_port = 30000
```
Comment (sushantdewan123) [inline README.md]: change distribution port to 2878

Comment (sushantdewan123) [reply README.md]: We recently switched from 40,000 to 2878

File: README.md [hunk 13]
```diff
@@ -1 +1,128 @@
...
+# wavefront_sender = Wavefront::WavefrontProxyClient.new(<proxy_host_ip>, metrics_port, distribution_port, tracing_port)
```
Comment (sushantdewan123) [inline README.md:67]: why is this commented out?

File: test/reporter_test.rb [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+require_relative '../lib/wavefrontopentracing/reporting/composite'
+require_relative '../lib/wavefrontopentracing/tracer'
+require_relative '../lib/wavefrontopentracing/span'
+
+class ReporterTest < Minitest::Test
+
+  def self.app_tags
+    Wavefront::ApplicationTags.new("wft2-shopping",
+                                   "shopping-service",
+                                   cluster: "virunga",
```
Comment (sushantdewan123) [inline test/reporter_test.rb]: dont put virunga in here
please put 
us-west-2 as an example

File: test/reporter_test.rb [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+                                   shard: "shopping-shard",
```
Comment (sushantdewan123) [inline test/reporter_test.rb]: use primary as an example for shard value

File: wavefront-opentracing-sdk-ruby.gemspec [hunk 16]
```diff
@@ -0,0 +1,33 @@
+lib = File.expand_path('lib', __dir__)
+$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
+require_relative 'lib/wavefrontopentracing/version'
+
+Gem::Specification.new do |spec|
+  spec.name = 'wavefront-opentracing-sdk'
+  spec.version = WavefrontOpentracing::VERSION
+  spec.authors = ['Gangadharaswamy']
+  spec.email = ['gangadhar@vmware.com']
```
Comment (sushantdewan123) [inline wavefront-opentracing-sdk-ruby.gemspec]: use the email address I had mentioned earlier

**Review by sushantdewan123 — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 11 above)
Comment (gangadharaswamy) [reply .travis.yml]: `wavefront-client` is for tier-1 SDK gem. For time-being it was commented. Will uncomment and check-in.

File: README.md (hunk 13 above)
Comment (gangadharaswamy) [reply README.md:67]: By default, direct ingestion sample code is uncommented. As an option ingestion via proxy code is commented out.

**Review by sushantdewan123 — APPROVED** [review]

**Pull request merged**